py-nominal-streaming/python/nominal_streaming/nominal_dataset_stream.py

```diff
@@ -0,0 +1,296 @@
+):
+    stream.enqueue("chanA", datetime.now(timezone.utc), 1.23, tags={"site": "a1"})
+    stream.enqueue_batch("chanB", [0, 1_000_000_000], [5, 6], tags={"phase": "prod"})
+    stream.enqueue_from_dict(0, {"chanC": "ok", "chanD": 7}, tags={"who": "tester"})
+
+"""
+
+from __future__ import annotations
+
+import datetime
+import logging
+import pathlib
+import signal
+from types import TracebackType
+from typing import Mapping, Sequence, Type
+
+from nominal_streaming._nominal_streaming import (
+    NominalStreamOpts,
+    _NominalDatasetStream,
+)
+
+logger = logging.getLogger(__name__)
+
+TimestampLike = str | int | datetime.datetime
+DataType = int | float | str
+
+
+class NominalDatasetStream:
+    """Top-level python wrapper for the Rust streaming client to Nominal."""
+
+    def __init__(self, auth_header: str, opts: NominalStreamOpts):
+        """Initializer for dataset stream.
+
+        Args:
+            auth_header: API key or access token to the Nominal API
+            opts: Optional options for the underlying stream
+        """
+        self._auth_header = auth_header
+        self._opts = opts
+        self._impl = _NominalDatasetStream(self._opts)
+        self._old_sigint = None
+
+    @classmethod
+    def from_settings(
+        cls,
+        auth_header: str,
+        api_base_url: str | None = None,
+        max_points_per_record: int | None = None,
+        max_request_delay: datetime.timedelta | None = None,
+        max_buffered_requests: int | None = None,
+        num_upload_workers: int | None = None,
+        num_runtime_workers: int | None = None,
+    ) -> NominalDatasetStream:
+        """Factory constructor to build a NominalDatasetStream using optional overrides for configuration options
+
+        Args:
+            auth_header: API Key or Personal Access Token for accessing the Nominal API
+            api_base_url: Overrides the default base API URL.
+            max_points_per_record: Overrides the default number of points that may be sent in a single batch
+            max_request_delay: Overrides the default maximum buffering time for data between flushes.
+                NOTE: if the amount of data being streamed is greater than available bandwidth, data may be
+                      buffered longer than the configured duration.
+            max_buffered_requests: Overrides the default number of requests that may be buffered between encoding
+                threads and upload threads. Increasing this may prevent blocking threads in situations with spotty
+                internet, but increase teardown time (e.g. when pressing ctrl + c)
+            num_upload_workers: Overrides the default number of upload worker threads
+                NOTE: should be set lower than the number of runtime workers.
+            num_runtime_workers: Overrides the default number of runtime worker threads
+                NOTE: should be set higher than the number of upload workers.
```

Comment:
Can we do a check for this `num_upload_workers < num_runtime_workers` constraint?

All of these say "override the default" and the default isn't specified or obvious from context. What if the `Opts` didn't have a default and Python defined the defaults?